rust-runtime/aws-smithy-runtime-api/src/client/interceptors.rs

```diff
@@ -679,48 +697,124 @@ impl Interceptors {
+        }
+        result.map_err(InterceptorError::read_before_execution)
+    }
+
+    interceptor_impl_fn!(mut modify_before_serialization);
+    interceptor_impl_fn!(ref read_before_serialization);
+    interceptor_impl_fn!(ref read_after_serialization);
+    interceptor_impl_fn!(mut modify_before_retry_loop);
+    interceptor_impl_fn!(ref read_before_attempt);
+    interceptor_impl_fn!(mut modify_before_signing);
+    interceptor_impl_fn!(ref read_before_signing);
+    interceptor_impl_fn!(ref read_after_signing);
+    interceptor_impl_fn!(mut modify_before_transmit);
+    interceptor_impl_fn!(ref read_before_transmit);
+    interceptor_impl_fn!(ref read_after_transmit);
+    interceptor_impl_fn!(mut modify_before_deserialization);
+    interceptor_impl_fn!(ref read_before_deserialization);
+    interceptor_impl_fn!(ref read_after_deserialization);
+
+    pub fn modify_before_attempt_completion(
+        &self,
+        ctx: &mut InterceptorContext<Input, Output, Error>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), InterceptorError> {
+        let mut result: Result<(), BoxError> = Ok(());
+        for interceptor in self.interceptors() {
+            if let Err(new_error) = interceptor.modify_before_attempt_completion(ctx, cfg) {
+                if let Err(last_error) = result {
+                    tracing::debug!("{}", DisplayErrorContext(&*last_error));
+                }
+                result = Err(new_error);
+            }
+        }
+        result.map_err(InterceptorError::modify_before_attempt_completion)
+    }
+
+    pub fn read_after_attempt(
+        &self,
+        ctx: &InterceptorContext<Input, Output, Error>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), InterceptorError> {
+        let mut result: Result<(), BoxError> = Ok(());
+        for interceptor in self.interceptors() {
+            if let Err(new_error) = interceptor.read_after_attempt(ctx, cfg) {
+                if let Err(last_error) = result {
+                    tracing::debug!("{}", DisplayErrorContext(&*last_error));
+                }
+                result = Err(new_error);
+            }
+        }
+        result.map_err(InterceptorError::read_after_attempt)
+    }
+
+    pub fn modify_before_completion(
+        &self,
+        ctx: &mut InterceptorContext<Input, Output, Error>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), InterceptorError> {
+        let mut result: Result<(), BoxError> = Ok(());
+        for interceptor in self.interceptors() {
+            if let Err(new_error) = interceptor.modify_before_completion(ctx, cfg) {
+                if let Err(last_error) = result {
+                    tracing::debug!("{}", DisplayErrorContext(&*last_error));
+                }
+                result = Err(new_error);
+            }
+        }
+        result.map_err(InterceptorError::modify_before_completion)
+    }
+
+    pub fn read_after_execution(
+        &self,
+        ctx: &InterceptorContext<Input, Output, Error>,
```

Comment:
Seems like the final hooks should also get a context wrapper to hide the internals, and that will make it possible to use the macro to make these.